andi/andi.py

```diff
@@ -1,62 +1,302 @@
+        for the input class/function can be resolved. When ``strict`` is False,
+        this function provides the plan only for those arguments that could
+        be resolved, so the last task of the plan could be incomplete.
+    :return: A plan
     """
-    if callable(arguments_or_func):
-        arguments = inspect(arguments_or_func)
-    else:
-        arguments = arguments_or_func
+    is_injectable = _ensure_can_provide_func(is_injectable)
+    externally_provided = _ensure_can_provide_func(externally_provided)
+
+    plan_odict = _plan(class_or_func,
+                 is_injectable=is_injectable,
+                 externally_provided=externally_provided,
+                 strict=strict,
+                 dependency_stack=None)
+    return Plan(plan_odict.items())
+
+
+_PlanDict = MutableMapping[Callable, Dict[str, Callable]]
+
+def _plan(class_or_func: Callable, *,
+          is_injectable: Callable[[Callable], bool],
+          externally_provided: Callable[[Callable], bool],
+          strict,
+          dependency_stack=None) -> _PlanDict:
+    dependency_stack = dependency_stack or []
+    plan_seq  = OrderedDict()  # type: _PlanDict
+    type_for_arg = {}
+
+    if externally_provided(class_or_func):
+        plan_seq[class_or_func] = {}
+        return plan_seq
 
-    if isinstance(can_provide, Container):
-        can_provide = can_provide.__contains__
+    if dependency_stack and not is_injectable(class_or_func):
+        raise NonProvidableError(
+            "Type {} cannot be provided".format(class_or_func))
+
+    if class_or_func in dependency_stack:
+        raise CyclicDependencyError(
+            "Cyclic dependency found. Dependency graph: {}".format(
+                " -> ".join(map(str, dependency_stack + [class_or_func]))))
+
+    dependency_stack = dependency_stack + [class_or_func]
+
+    if isinstance(class_or_func, type):
+        cls = cast(Type, class_or_func)
+        arguments = inspect(cls.__init__)
+    else:
+        arguments = inspect(class_or_func)
 
-    result = {}
     for argname, types in arguments.items():
-        for cls in types:
-            if can_provide(cls):
-                result[argname] = cls
-                break
-    return result
+        sel_cls = _select_type(types, is_injectable, externally_provided)
+        if sel_cls is not None:
+            if sel_cls not in plan_seq:
+                plan = _plan(sel_cls,
+                             is_injectable=is_injectable,
+                             externally_provided=externally_provided,
+                             strict=True,
+                             dependency_stack=dependency_stack)
+                plan_seq.update(plan)
+            type_for_arg[argname] = sel_cls
+        else:
+            if strict:
+                if not types:
+                    msg = "Parameter '{}' is lacking annotations in " \
+                          "'{}.__init__()'. Not possible to build a plan".format(
```

Comment:
Why can we guarantee that class_or_func is a class here?

Reply:
Ops, great catch. This got wrong in a refactoring. Fixed at https://github.com/scrapinghub/andi/pull/8/commits/d5cfcd1ae6475ebebb86eb565ff3c57360b773c2